File: bindings/core/src/method_handler/account.rs
// Copyright 2023 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

use iota_sdk::{
    client::api::{
        PreparedTransactionData, PreparedTransactionDataDto, SignedTransactionData, SignedTransactionDataDto,
    },
    types::{
        block::output::{dto::OutputDto, Output},
        TryFromDto,
    },
    wallet::account::{types::TransactionDto, OutputDataDto, PreparedCreateNativeTokenTransactionDto},
};

use crate::{method::WalletMethod, Response, Result};

pub(crate) async fn call_account_method_internal(account: &Account, method: WalletMethod) -> Result<Response> {
    let response = match method {
        WalletMethod::Addresses => {
            let addresses = account.addresses().await?;
            Response::Addresses(addresses)
        }
        WalletMethod::AddressesWithUnspentOutputs => {
            let addresses = account.addresses_with_unspent_outputs().await?;
            Response::AddressesWithUnspentOutputs(addresses)
        }
        WalletMethod::ClaimableOutputs { outputs_to_claim } => {
            let output_ids = account.claimable_outputs(outputs_to_claim).await?;
            Response::OutputIds(output_ids)
        }
        WalletMethod::ClaimOutputs { output_ids_to_claim } => {
            let transaction = account.claim_outputs(output_ids_to_claim.to_vec()).await?;
            Response::SentTransaction(TransactionDto::from(&transaction))
        }
        #[cfg(feature = "participation")]
        WalletMethod::DeregisterParticipationEvent { event_id } => {
            account.deregister_participation_event(&event_id).await?;
            Response::Ok
        }
        WalletMethod::GenerateEd25519Addresses { amount, options } => {
            let address = account.generate_ed25519_addresses(amount, options).await?;
            Response::GeneratedAccountAddresses(address)
        }
        WalletMethod::GetBalance => Response::Balance(account.balance().await?),
        WalletMethod::GetFoundryOutput { token_id } => {
            let output = account.get_foundry_output(token_id).await?;
            Response::Output(OutputDto::from(&output))
        }
        WalletMethod::GetIncomingTransaction { transaction_id } => {
            let transaction = account.get_incoming_transaction(&transaction_id).await;

            transaction.map_or_else(
                || Response::Transaction(None),
                |transaction| Response::Transaction(Some(Box::new(TransactionDto::from(&transaction)))),
            )
        }
        WalletMethod::GetOutput { output_id } => {
            let output_data = account.get_output(&output_id).await;
            Response::OutputData(output_data.as_ref().map(OutputDataDto::from).map(Box::new))
        }
        #[cfg(feature = "participation")]
        WalletMethod::GetParticipationEvent { event_id } => {
            let event_and_nodes = account.get_participation_event(event_id).await?;
            Response::ParticipationEvent(event_and_nodes)
        }
        #[cfg(feature = "participation")]
        WalletMethod::GetParticipationEventIds { node, event_type } => {
            let event_ids = account.get_participation_event_ids(&node, event_type).await?;
            Response::ParticipationEventIds(event_ids)
        }
        #[cfg(feature = "participation")]
        WalletMethod::GetParticipationEventStatus { event_id } => {
            let event_status = account.get_participation_event_status(&event_id).await?;
            Response::ParticipationEventStatus(event_status)
        }
        #[cfg(feature = "participation")]
        WalletMethod::GetParticipationEvents => {
            let events = account.get_participation_events().await?;
            Response::ParticipationEvents(events)
        }
        #[cfg(feature = "participation")]
        WalletMethod::GetParticipationOverview { event_ids } => {
            let overview = account.get_participation_overview(event_ids).await?;
            Response::AccountParticipationOverview(overview)
        }
        WalletMethod::GetTransaction { transaction_id } => {
            let transaction = account.get_transaction(&transaction_id).await;
            Response::Transaction(transaction.as_ref().map(TransactionDto::from).map(Box::new))
        }
        #[cfg(feature = "participation")]
        WalletMethod::GetVotingPower => {
            let voting_power = account.get_voting_power().await?;
            Response::VotingPower(voting_power.to_string())
        }
        WalletMethod::IncomingTransactions => {
            let transactions = account.incoming_transactions().await;
            Response::Transactions(transactions.iter().map(TransactionDto::from).collect())
        }
        WalletMethod::Outputs { filter_options } => {
            let outputs = account.outputs(filter_options).await?;
            Response::OutputsData(outputs.iter().map(OutputDataDto::from).collect())
        }
        WalletMethod::PendingTransactions => {
            let transactions = account.pending_transactions().await;
            Response::Transactions(transactions.iter().map(TransactionDto::from).collect())
        }
        WalletMethod::PrepareBurn { burn, options } => {
            let data = account.prepare_burn(burn, options).await?;
            Response::PreparedTransaction(PreparedTransactionDataDto::from(&data))
        }
        WalletMethod::PrepareConsolidateOutputs { params } => {
            let data = account.prepare_consolidate_outputs(params).await?;
            Response::PreparedTransaction(PreparedTransactionDataDto::from(&data))
        }
        WalletMethod::PrepareCreateAccountOutput { params, options } => {
            let data = account.prepare_create_account_output(params, options).await?;
            Response::PreparedTransaction(PreparedTransactionDataDto::from(&data))
        }
        WalletMethod::PrepareMeltNativeToken {
            token_id,
            melt_amount,
            options,
        } => {
            let data = account
                .prepare_melt_native_token(token_id, melt_amount, options)
                .await?;
            Response::PreparedTransaction(PreparedTransactionDataDto::from(&data))
        }
        #[cfg(feature = "participation")]
        WalletMethod::PrepareDecreaseVotingPower { amount } => {
            let data = account.prepare_decrease_voting_power(amount).await?;
            Response::PreparedTransaction(PreparedTransactionDataDto::from(&data))
        }
        WalletMethod::PrepareMintNativeToken {
            token_id,
            mint_amount,
            options,
        } => {
            let data = account
                .prepare_mint_native_token(token_id, mint_amount, options)
                .await?;
            Response::PreparedTransaction(PreparedTransactionDataDto::from(&data))
        }
        #[cfg(feature = "participation")]
        WalletMethod::PrepareIncreaseVotingPower { amount } => {
            let data = account.prepare_increase_voting_power(amount).await?;
            Response::PreparedTransaction(PreparedTransactionDataDto::from(&data))
        }
        WalletMethod::PrepareMintNfts { params, options } => {
            let data = account.prepare_mint_nfts(params, options).await?;
            Response::PreparedTransaction(PreparedTransactionDataDto::from(&data))
        }
        WalletMethod::PrepareCreateNativeToken { params, options } => {
            let data = account.prepare_create_native_token(params, options).await?;
            Response::PreparedCreateNativeTokenTransaction(PreparedCreateNativeTokenTransactionDto::from(&data))
        }
        WalletMethod::PrepareOutput {
            params,
            transaction_options,
        } => {
            let output = account.prepare_output(*params, transaction_options).await?;
            Response::Output(OutputDto::from(&output))
        }
        WalletMethod::PrepareSend { params, options } => {
            let data = account.prepare_send(params, options).await?;
            Response::PreparedTransaction(PreparedTransactionDataDto::from(&data))
        }
        WalletMethod::PrepareSendNativeTokens { params, options } => {
            let data = account.prepare_send_native_tokens(params.clone(), options).await?;
            Response::PreparedTransaction(PreparedTransactionDataDto::from(&data))
        }
        WalletMethod::PrepareSendNft { params, options } => {
            let data = account.prepare_send_nft(params.clone(), options).await?;
            Response::PreparedTransaction(PreparedTransactionDataDto::from(&data))
        }
        #[cfg(feature = "participation")]
        WalletMethod::PrepareStopParticipating { event_id } => {
            let data = account.prepare_stop_participating(event_id).await?;
            Response::PreparedTransaction(PreparedTransactionDataDto::from(&data))
        }
        WalletMethod::PrepareTransaction { outputs, options } => {
            let token_supply = account.client().get_token_supply().await?;
            let data = account
                .prepare_transaction(
                    outputs
                        .into_iter()
                        .map(|o| Ok(Output::try_from_dto_with_params(o, token_supply)?))
                        .collect::<Result<Vec<Output>>>()?,
                    options,
                )
                .await?;
            Response::PreparedTransaction(PreparedTransactionDataDto::from(&data))
        }
        #[cfg(feature = "participation")]
        WalletMethod::PrepareVote { event_id, answers } => {
            let data = account.prepare_vote(event_id, answers).await?;
            Response::PreparedTransaction(PreparedTransactionDataDto::from(&data))
        }
        #[cfg(feature = "participation")]
        WalletMethod::RegisterParticipationEvents { options } => {
            let events = account.register_participation_events(&options).await?;
            Response::ParticipationEvents(events)
        }
        WalletMethod::ReissueTransactionUntilIncluded {
            transaction_id,
            interval,
            max_attempts,
        } => {
            let block_id = account
                .reissue_transaction_until_included(&transaction_id, interval, max_attempts)
                .await?;
            Response::BlockId(block_id)
        }
        WalletMethod::Send {
            amount,
            address,
            options,
        } => {
            let transaction = account.send(amount, address, options).await?;
            Response::SentTransaction(TransactionDto::from(&transaction))
        }
        WalletMethod::SendWithParams { params, options } => {
            let transaction = account.send_with_params(params, options).await?;
            Response::SentTransaction(TransactionDto::from(&transaction))
        }
        WalletMethod::SendOutputs { outputs, options } => {
            let token_supply = account.client().get_token_supply().await?;
            let transaction = account
                .send_outputs(
                    outputs
                        .into_iter()
                        .map(|o| Ok(Output::try_from_dto_with_params(o, token_supply)?))
                        .collect::<iota_sdk::wallet::Result<Vec<Output>>>()?,
                    options,
                )
                .await?;
            Response::SentTransaction(TransactionDto::from(&transaction))
        }
        WalletMethod::SetAlias { alias } => {
            account.set_alias(&alias).await?;
            Response::Ok
        }
        WalletMethod::SetDefaultSyncOptions { options } => {
            account.set_default_sync_options(options).await?;
            Response::Ok
        }
        WalletMethod::SignAndSubmitTransaction {
            prepared_transaction_data,
        } => {
            let transaction = account
                .sign_and_submit_transaction(
                    PreparedTransactionData::try_from_dto_with_params(
                        prepared_transaction_data,
                        account.client().get_protocol_parameters().await?,
                    )?,
                    None,
                )
                .await?;
            Response::SentTransaction(TransactionDto::from(&transaction))
        }
        WalletMethod::SignTransactionEssence {
            prepared_transaction_data,
        } => {
            let signed_transaction_data = account
                .sign_transaction_essence(&PreparedTransactionData::try_from_dto(prepared_transaction_data)?)
                .await?;
            Response::SignedTransactionData(SignedTransactionDataDto::from(&signed_transaction_data))
        }
        WalletMethod::SubmitAndStoreTransaction {
            signed_transaction_data,
        } => {
            let signed_transaction_data = SignedTransactionData::try_from_dto_with_params(
                signed_transaction_data,
                account.client().get_protocol_parameters().await?,
            )?;
            let transaction = account
                .submit_and_store_transaction(signed_transaction_data, None)
                .await?;
            Response::SentTransaction(TransactionDto::from(&transaction))
        }
        WalletMethod::Sync { options } => Response::Balance(account.sync(options).await?),
        WalletMethod::Transactions => {
            let transactions = account.transactions().await;
            Response::Transactions(transactions.iter().map(TransactionDto::from).collect())
        }
        WalletMethod::UnspentOutputs { filter_options } => {
            let outputs = account.unspent_outputs(filter_options).await?;
            Response::OutputsData(outputs.iter().map(OutputDataDto::from).collect())
        }
    };
    Ok(response)
}

File: bindings/nodejs-old/types/bridge/account.ts
import type {
    OutputTypes,
    HexEncodedAmount,
    HexEncodedString,
} from '@iota/types';
import type { SyncOptions, FilterOptions } from '../account';
import type {
    SendParams,
    SendNativeTokensParams,
    SendNftParams,
    GenerateAddressOptions,
    GenerateAddressesOptions,
} from '../address';
import type {
    BuildAliasOutputData,
    BuildBasicOutputData,
    BuildFoundryOutputData,
    BuildNftOutputData,
} from '../buildOutputData';
import type { Node } from '../network';
import type { OutputParams } from '../outputParams';
import type { OutputsToClaim } from '../output';
import type { SignedTransactionEssence } from '../signedTransactionEssence';
import type { PreparedTransactionData } from '../preparedTransactionData';
import type {
    AliasOutputParams,
    CreateNativeTokenParams,
    TransactionOptions,
    MintNftParams,
} from '../transactionOptions';
import type {
    ParticipationEventId,
    ParticipationEventRegistrationOptions,
    ParticipationEventType,
} from '../participation';
import type { ConsolidationParams } from '../consolidationParams';
import { Bip44, Ed25519Signature } from '../secretManager';

export type __BuildAliasOutputMethod__ = {
    name: 'buildAliasOutput';
    data: BuildAliasOutputData;
};

export type __BuildBasicOutputMethod__ = {
    name: 'buildBasicOutput';
    data: BuildBasicOutputData;
};

export type __BuildFoundryOutputMethod__ = {
    name: 'buildFoundryOutput';
    data: BuildFoundryOutputData;
};

export type __BuildNftOutputMethod__ = {
    name: 'buildNftOutput';
    data: BuildNftOutputData;
};

export type __BurnNativeTokenMethod__ = {
    name: 'burnNativeToken';
    data: {
        tokenId: string;
        burnAmount: HexEncodedAmount;
        options?: TransactionOptions;
    };
};

export type __BurnNftMethod__ = {
    name: 'burnNft';
    data: {
        nftId: string;
        options?: TransactionOptions;
    };
};

export type __ClaimOutputsMethod__ = {
    name: 'claimOutputs';
    data: {
        outputIdsToClaim: string[];
    };
};

export type __ConsolidateOutputsMethod__ = {
    name: 'consolidateOutputs';
    data: {
        params: ConsolidationParams;
    };
};

export type __CreateAliasOutputMethod__ = {
    name: 'createAliasOutput';
    data: {
        params?: AliasOutputParams;
        options?: TransactionOptions;
    };
};

export type __MeltNativeTokenMethod__ = {
    name: 'meltNativeToken';
    data: {
        tokenId: string;
        meltAmount: HexEncodedAmount;
        options?: TransactionOptions;
    };
};

export type __DeregisterParticipationEventMethod__ = {
    name: 'deregisterParticipationEvent';
    data: {
        eventId: ParticipationEventId;
    };
};

export type __DestroyAliasMethod__ = {
    name: 'destroyAlias';
    data: {
        aliasId: string;
        options?: TransactionOptions;
    };
};

export type __DestroyFoundryMethod__ = {
    name: 'destroyFoundry';
    data: {
        foundryId: string;
        options?: TransactionOptions;
    };
};

export type __GenerateEd25519AddressesMethod__ = {
    name: 'generateEd25519Addresses';
    data: {
        amount: number;
        options?: GenerateAddressOptions;
    };
};

export type __GenerateEvmAddressesMethod__ = {
    name: 'generateEvmAddresses';
    data: {
        options: GenerateAddressesOptions;
    };
};

export type __VerifyEd25519SignatureMethod__ = {
    name: 'verifyEd25519Signature';
    data: {
        signature: Ed25519Signature;
        message: HexEncodedString;
    };
};

export type __VerifySecp256k1EcdsaSignatureMethod__ = {
    name: 'verifySecp256k1EcdsaSignature';
    data: {
        publicKey: HexEncodedString;
        signature: HexEncodedString;
        message: HexEncodedString;
    };
};

export type __SignSecp256k1EcdsaMethod__ = {
    name: 'signSecp256k1Ecdsa';
    data: {
        message: HexEncodedString;
        chain: Bip44;
    };
};

export type __GetBalanceMethod__ = {
    name: 'getBalance';
};

export type __GetIncomingTransactionMethod__ = {
    name: 'getIncomingTransaction';
    data: {
        transactionId: string;
    };
};

export type __GetOutputMethod__ = {
    name: 'getOutput';
    data: {
        outputId: string;
    };
};

export type __GetFoundryOutputMethod__ = {
    name: 'getFoundryOutput';
    data: {
        tokenId: string;
    };
};

export type __ClaimableOutputsMethod__ = {
    name: 'claimableOutputs';
    data: {
        outputsToClaim: OutputsToClaim;
    };
};

export type __GetTransactionMethod__ = {
    name: 'getTransaction';
    data: {
        transactionId: string;
    };
};

export type __AddressesMethod__ = {
    name: 'addresses';
};

export type __AddressesWithUnspentOutputsMethod__ = {
    name: 'addressesWithUnspentOutputs';
};

export type __OutputsMethod__ = {
    name: 'outputs';
    data: {
        filterOptions?: FilterOptions;
    };
};

export type __PendingTransactionsMethod__ = {
    name: 'pendingTransactions';
};

export type __IncomingTransactionsMethod__ = {
    name: 'incomingTransactions';
};

export type __TransactionsMethod__ = {
    name: 'transactions';
};

export type __UnspentOutputsMethod__ = {
    name: 'unspentOutputs';
    data: {
        filterOptions?: FilterOptions;
    };
};

export type __MinimumRequiredStorageDepositMethod__ = {
    name: 'minimumRequiredStorageDeposit';
    data: {
        output: OutputTypes;
    };
};

export type __MintNativeTokenMethod__ = {
    name: 'mintNativeToken';
    data: {
        tokenId: string;
        mintAmount: HexEncodedAmount;
        options?: TransactionOptions;
    };
};

export type __CreateNativeTokenMethod__ = {
    name: 'createNativeToken';
    data: {
        params: CreateNativeTokenParams;
        options?: TransactionOptions;
    };
};

export type __MintNftsMethod__ = {
    name: 'mintNfts';
    data: {
        params: MintNftParams[];
        options?: TransactionOptions;
    };
};

export type __PrepareOutputMethod__ = {
    name: 'prepareOutput';
    data: {
        params: OutputParams;
        transactionOptions?: TransactionOptions;
    };
};

export type __PrepareSendMethod__ = {
    name: 'prepareSend';
    data: {
        params: SendParams[];
        options?: TransactionOptions;
    };
};

export type __PrepareTransactionMethod__ = {
    name: 'prepareTransaction';
    data: {
        outputs: OutputTypes[];
        options?: TransactionOptions;
    };
};

export type __RegisterParticipationEventsMethod__ = {
    name: 'registerParticipationEvents';
    data: {
        options: ParticipationEventRegistrationOptions;
    };
};

export type __RequestFundsFromFaucetMethod__ = {
    name: 'requestFundsFromFaucet';
    data: {
        url: string;
        address: string;
    };
};

export type __RetryTransactionUntilIncludedMethod__ = {
    name: 'retryTransactionUntilIncluded';
    data: {
        transactionId: string;
        interval?: number;
        maxAttempts?: number;
    };
};

export type __SendMethod__ = {
    name: 'send';
    data: {
        params: SendParams[];
        options?: TransactionOptions;
    };
};

export type __SendNativeTokensMethod__ = {
    name: 'sendNativeTokens';
    data: {
        params: SendNativeTokensParams[];
        options?: TransactionOptions;
    };
};

export type __SendNftMethod__ = {
    name: 'sendNft';
    data: {
        params: SendNftParams[];
        options?: TransactionOptions;
    };
};

export type __SendOutputsMethod__ = {
    name: 'sendOutputs';
    data: {
        outputs: OutputTypes[];
        options?: TransactionOptions;
    };
};

export type __SetAliasMethod__ = {
    name: 'setAlias';
    data: {
        alias: string;
    };
};

export type __SetDefaultSyncOptionsMethod__ = {
    name: 'setDefaultSyncOptions';
    data: {
        options: SyncOptions;
    };
};

export type __SignTransactionEssenceMethod__ = {
    name: 'signTransactionEssence';
    data: {
        preparedTransactionData: PreparedTransactionData;
    };
};

export type __SubmitAndStoreTransactionMethod__ = {
    name: 'submitAndStoreTransaction';
    data: {
        signedTransactionData: SignedTransactionEssence;
    };
};

export type __SyncAccountMethod__ = {
    name: 'syncAccount';
    data: {
        options?: SyncOptions;
    };
};

export type __VoteMethod__ = {
    name: 'vote';
    data: {
        eventId?: ParticipationEventId;
        answers?: number[];
    };
};

export type __StopParticipatingMethod__ = {
    name: 'stopParticipating';
    data: {
        eventId: ParticipationEventId;
    };
};

export type __GetParticipationOverviewMethod__ = {
    name: 'getParticipationOverview';
    data: {
        eventIds?: ParticipationEventId[];
    };
};

export type __IncreaseVotingPowerMethod__ = {
    name: 'increaseVotingPower';
    data: {
        amount: string;
    };
};

export type __GetParticipationEventMethod__ = {
    name: 'getParticipationEvent';
    data: {
        eventId: ParticipationEventId;
    };
};

export type __GetParticipationEventIdsMethod__ = {
    name: 'getParticipationEventIds';
    data: {
        node: Node;
        eventType?: ParticipationEventType;
    };
};

export type __GetParticipationEventsMethod__ = {
    name: 'getParticipationEvents';
};

export type __GetParticipationEventStatusMethod__ = {
    name: 'getParticipationEventStatus';
    data: {
        eventId: ParticipationEventId;
    };
};

export type __DecreaseVotingPowerMethod__ = {
    name: 'decreaseVotingPower';
    data: {
        amount: string;
    };
};
